### DISCUSSION TITLE
substr instruction

### DISCUSSION THREAD

**Discussion by Peter-Jacob** [adesutherland/CREXX#112] [Q&A]
I am experimenting with a substr instrution, 

 substr  target-reg,source-reg,offset-reg   for the length-reg I had the idea to use the next register following offset-reg. We do something similar with call, I looked at this coding but couldn't find the trick.

so what I want is:
```
   load r1,"1234567890"
   load r3,5
   load r4,3
   substr r2,r1,r3  
```
and fetch internally r4 as length register.

**Comment by adesutherland** [comment]
The way I was envisaging it was with setstrpos 
```
   load r1,"1234567890"
   setstrpos r1,5  * Start pos
   load r4,3       * length
   substr r2,r1,r4 * Result in r2
```

I think this will be as fast as your scheme (well maybe!)

For your solution:

In rxvmintp.c lines from 482 (note the comment!!), this reminds me that the "past is a foreign country, they do strange things there", anyway I can't remember writing this code!
```
            /* Arguments - complex lets never have to change this code! */
            current_frame->locals[current_frame->module->globals + p2->locals] =
                    current_frame->parent->locals[(pc + (3))->index];
            for (i = 0; i < v3->int_value; i++) {
                current_frame->locals[current_frame->module->globals +
                                      p2->locals + i + 1] =
                        current_frame->parent->locals[(pc + (3))->index + i +
                                                      1];
            }
```

So I think the incantation you are looking for is (untested!!!)

Argument 3 register is this (this is macro op3R)
```
    current_frame->locals[(pc + 3)->index]
```
So the register after the register in argument 3 (what you want) is
```
    current_frame->locals[ (pc + 3)->index + 1 ];  /* The 3 means the 3rd argument, the +1 gets the next register up */
```
Good luck!!

**Reply by Peter-Jacob** [reply]
Thank you! 
"Past is a foreign country", I know thousands of foreign counties  ;)

**Reply by Peter-Jacob** [reply]
first cut of a substr instruction added (f0030), volunteers needed!
There is no checking, this must be done in the calling assembler program.

**Reply by adesutherland** [reply]
Something I am thinking though is how to make the REXX Level B assembler command reserve registers (so you can have two consecutive ones)

**Comment by Peter-Jacob** [comment]
Before I start to add the chars to the target string I need to clear it. I use 
 `  v1->string_length=0;`
But I am not happy with it as, the original string may be longer, so we may end up with allocated unused storage. Is there a more secure way to drop the entire string, if there was one?

**Reply by adesutherland** [reply]
Remember to set to zero: 
    string_pos 

and (#ifndef NUTF8)
    string_chars;
    string_char_pos;

They all should be zero.

Then when you put in the new string you need to ensure that string_length and string_chars are right (hopefully from counting during the copy).

Apologies if I am teaching you to suck eggs (if that metaphor translates)!

Anyway as far as freeing memory - I think the best answer (certainly at the moment) is don't worry about it. I suspect that in real life it will never be an issue. And if it is we will have to have some kind of garbage collector on another thread - but I honestly think "why bother" ...
